cranelift/codegen/src/isa/x86/abi.rs

```diff
@@ -525,50 +523,60 @@ fn fastcall_prologue_epilogue(func: &mut ir::Function, isa: &dyn TargetIsa) -> C
         panic!("TODO: windows-fastcall: x86-32 not implemented yet");
     }
 
-    let csrs = callee_saved_regs_used(isa, func);
-
     // The reserved stack area is composed of:
-    //   return address + frame pointer + all callee-saved registers + shadow space
+    //   return address + frame pointer + all callee-saved registers
     //
     // Pushing the return address is an implicit function of the `call`
     // instruction. Each of the others we will then push explicitly. Then we
     // will adjust the stack pointer to make room for the rest of the required
     // space for this frame.
-    let word_size = isa.pointer_bytes() as usize;
-    let num_fprs = csrs.iter(FPR).len();
-    let csr_stack_size = ((csrs.iter(GPR).len() + 2) * word_size) as i32;
+    let csrs = callee_saved_regs_used(isa, func);
+    let gpsr_stack_size = ((csrs.iter(GPR).len() + 2) * isa.pointer_bytes() as usize) as u32;
+    let fpsr_stack_size = (csrs.iter(FPR).len() * types::F64X2.bytes() as usize) as u32;
+    let mut csr_stack_size = gpsr_stack_size + fpsr_stack_size;
 
-    // Only create an FPR stack slot if we're going to save FPRs.
-    let fpr_slot = if num_fprs > 0 {
-        // Create a stack slot for FPRs to be preserved in. This is an `ExplicitSlot` because it
-        // seems to most closely map to it as a `StackSlotKind`: FPR preserve/restore should be
-        // through `stack_load` and `stack_store` (see later comment about issue #1198). Even
-        // though in a certain light FPR preserve/restore is "spilling" an argument, regalloc
-        // implies that `SpillSlot` may be eligible for certain optimizations, and we know with
-        // certainty that this space may not be reused in the function, nor moved around.
-        Some(func.create_stack_slot(ir::StackSlotData {
-            kind: ir::StackSlotKind::ExplicitSlot,
-            size: (num_fprs * types::F64X2.bytes() as usize) as u32,
-            offset: None,
-        }))
-    } else {
-        None
-    };
+    // FPRs must be saved with 16-byte alignment; because they follow the GPRs on the stack, align if needed
+    if fpsr_stack_size > 0 {
+        csr_stack_size += gpsr_stack_size & isa.pointer_bytes() as u32;
```

Comment:
To be correct for cases where `isa.pointer_bytes() != 8`:
```suggestion
        csr_stack_size = (csr_stack_size + 15) & !15;
```

Reply:
I think that's clearer and congruent with other places where we round up for alignment.  I'll change.